_posts/2021-05-28-why-avoid-an-orm.md

```diff
@@ -0,0 +1,231 @@
+Who's going to be able to spot this kind of problem in an actual code base with actual business logic everywhere? Who even knows if your particular ORM happens to work exactly like this?
+Not me.
+ 
+ 
+ 
+ 
+## Does this all sound overly complex?
+ 
+That's because it is. There's a simple solution not to solve these problems, but to get rid of them completely: **Don't use an ORM**.
+
+You can still use an ORM tool as long as you just avoid its ORM parts. For example, your ORM probably offers a DSL to construct type-safe queries ([LINQ](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/) or [JPA Criteria API](https://en.wikibooks.org/wiki/Java_Persistence/Criteria) for example). Using it doesn't yet create these problems because it's about constructing a specific query returning a result, and the objects don't need to have anything to do with ORMs.
+ 
+In my current project, we are using Java Hibernate, which is an ORM. Our entities don't have any methods to return other entities. Our department wouldn't have a `getEmployees()` method. We can do the same thing with a thin query library on top of the Hibernate/JPA, so that this:
+```java
+List<Employee> emps = department.getEmployees();
+```
+would look like this:
+```java
+Collection<Employee> emps = dao.getMany(query.related(department, Department_.employees));
+```
+It makes performing queries just as easy, but explicit. And it completely avoids both n+1 and relationship problems. This is even better typed since it returns a Collections instead of an (ordered) List since we didn't give an explicit ordering for the query.
+ 
+We can just as easily do projections instead of fetching the whole employees:
+```java
+Collection<String> emps = dao.getMany(query.related(department, Department_.employees), Project.value(Employee_.name));
+```
+because we are just constructing a query and not trying to interpret an object graph. Easy projections are still - as far as I know - an unsolved problem in ORMs.
+ 
+We can use `Id<Department>` for referencing the identity everywhere since we aren't pretending to be dealing with an object graph, except we still need to fetch the actual entity for updates - like `employee.setDepartment(actualDepartmentOrProxy)` - since that's how Hibernate works. But single-table updates are easier by modifying an object instead of creating SQL updates, and most of our entities are immutable anyway, so I can live with this tradeoff.
+ 
+ 
+ 
+## So my advice,
+
+if You decide to take it into account:
+- don't use an ORM
+  - In my current Java project, we use Hibernate, but something like [jOOQ](http://www.jooq.org) would probably be a lot better, and a lot less of an ORM.
+- think database first
+  - design your database in SQL, and make your Java/C#/etc tool understand that. Not the other way round. This also applies to migrations where (in Java) something like [DBMaintain](http://www.dbmaintain.org) or [Flyway](https://flywaydb.org) are nice tools.
```

Comment:
This suggestion doesn't really follow from the above problems. Something like a link to some other post with the reasoning behind this might be useful? (The links at the start somewhat cover this subject, though I'm not sure I buy those arguments.)

Reply:
You are right, but I wouldn't want to remove it since I think it's an important suggestion.
I found a nice link to go with it, do you think it's better now?

Reply:
The link definitely improves it! (Good reading too.)